### ISSUE TITLE
Unable to connect to AWS RDS database Moodle ECS cluster

### ISSUE THREAD

**Issue by adam-phillipps** [bitnami/bitnami-docker-moodle#75]
**Description**

I am unable to get the `bitnami/moodle` Docker container to connect to an AWS RDS database.
The database has the correct configuration, as far as I can make out and I am able to connect to it from the Moodle container but Moodle itself always dies.  It looks like it is having problems when Moodle is initializing, during `nami_initialize moodle` but I could be wrong about that.

The ECS cluster has Moodle running with the standard 80 and 443 ports open, running in a `Bridge` network that has access to the database cluster.  I'm pretty sure the AWS side of the equation is set up correctly but the way I'm configuring Moodle must not be correct, is my guess. 

Spinning up Moodle and the Maria DB image from Bitnami is as easy as a `docker-compose up` command but when I attempt to swap out Maria for the connection information to the remote database I get the same issue; The bootup script(s) seem to run fine until it works on the Moodle install/initialize/startup process.

**Steps to reproduce the issue:**
Run locally(L) or on an ECS cluster(C)
1. `$ curl -sSL https://raw.githubusercontent.com/bitnami/bitnami-docker-moodle/master/docker-compose.yml > docker-compose.yml`
2. 
	L. Comment out the Maria container and replace the values for the environment variables with the configuration for your RDS cluster, in step
	C. Create an ECS cluster that has a security group that can access the RDS cluster you build 
3. Create an Aurora RDS cluster that uses MySQL for its engine and that lives in the same security group as the ECS cluster.
4. 
	L. Replace the config in the `docker-compose.yml` file from step 1 with the values you get from your database cluster in step 2.
	C. Create an ECS Task that uses the `bitnami/moodle` image and has all the environment variables filled out.  It needs a `Volume` which I gave a host path of `/usr/local/bitnami` and the container path is `/bitnami`.
5. 
	L. Run `docker-compose up`
	C. Create an ECS Service that runs the task from step 3.  This will start it up automatically and re-run it, if it's failing.
6. Watch the logs...

**Describe the results you received:**

The task is started up and successfully initializes Apache, PHP and MySQL.  It then tries to initialize Moodle.  It looks like the first steps for the Moodle script are to set up the database because the next lines in the log are `mysql-c` related.  After what appears to be a successful authentication process, it times out and gives an error message that reads `Error executing 'postInstallation': Program exited with exit code 1`

In the database dashboard, I can see that it does look like Moodle is _trying_ to connect but it always fails.  The reason I think that is because I can see the right number of login attempts, about 1 per minute because it times out after that long and tries again, and because I see them at the right times of the day, like when I'm working on it.  
It's not definitive proof that Moodle is attempting to login but I'd be surprised to find out it's something else.
I haven't figured out how to find out why the login attempts are failing either.  
I've enabled better logging in MySQL/RDS to try to track down the issue, to no avail.  I can't even find the login attempts that are failing, whether they're Moodle or not.  I can see me connecting from my local machine, the host in the cluster and the container in the cluster.  Then there's a lot of backend looking MySQL queries, so no useful clues there, for me.

Here's the logs that I can see:
```
Welcome to the Bitnami moodle container
Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-moodle
Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-moodle/issues

nami    INFO  Initializing apache
apache  INFO  ==> Reconfiguring PID file location...
nami    INFO  apache successfully initialized
nami    INFO  Initializing php
nami    INFO  php successfully initialized
nami    INFO  Initializing mysql-client
nami    INFO  mysql-client successfully initialized
nami    INFO  Initializing moodle
mysql-c INFO  Trying to connect to MySQL server
mysql-c INFO Found MySQL server listening at <my-database-cluster>:3306
mysql-c INFO MySQL server listening and working at <my-database-cluster>:3306
Error executing 'postInstallation': Program exited with exit code 1
```

**Describe the results you expected:**

The Moodle container, whether running locally or in the cluster, can startup and run using a remote database instead of the MariaDB container.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Version**

- Output of `docker version`:

```
$ docker version
Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.4
 Git commit:   3dfb8343b139d6342acfd9975d7f1068b5b1c3d3
 Built:        Thu May 24 22:21:27 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.4
  Git commit:   7390fc6/18.03.1-ce
  Built:        Thu May 24 22:22:43 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

- Output of `docker info`:

```
$ docker info
Containers: 97
 Running: 2
 Paused: 0
 Stopped: 95
Images: 3
Server Version: 18.03.1-ce
Storage Driver: devicemapper
 Pool Name: docker-docker--pool
 Pool Blocksize: 524.3kB
 Base Device Size: 10.74GB
 Backing Filesystem: ext4
 Udev Sync Supported: true
 Data Space Used: 3.836GB
 Data Space Total: 23.33GB
 Data Space Available: 19.49GB
 Metadata Space Used: 4.997MB
 Metadata Space Total: 41.94MB
 Metadata Space Available: 36.95MB
 Thin Pool Minimum Free Space: 2.333GB
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.135-RHEL7 (2016-11-16)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.14.33-51.37.amzn1.x86_64
Operating System: Amazon Linux AMI 2018.03
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.792GiB
Name: ip-172-31-20-173
ID: FG2J:ZD2O:TRQ6:T6AV:WPWT:CWAD:U2O6:BOWJ:3C53:NEU6:5HGR:THL7
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

- Output of `docker-compose version` (if applicable):

```
# on my local machine
$ docker-compose version
docker-compose version 1.21.1, build 5a3f1a3
docker-py version: 3.3.0
CPython version: 3.6.4
OpenSSL version: OpenSSL 1.0.2o  27 Mar 2018
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**
- Entrypoint: `["/app-entrypoint.sh"]`
- CMD: `["/run.sh"]`
- AMI: `amzn-ami-2018.03.a-amazon-ecs-optimized (ami-d2f489aa)`
- Instance type: `t2-large`:
	- cores: 2
	- Memory: 8
	- AMI/VM: Amazon Linux AMI (ECS optimized)

I'm not 100% certain how ECS runs the `docker run` command but here's what the container looks like, as it's running:
environment variables I pass in at runtime:
- Environment:
```
ALLOW_EMPTY_PASSWORD=no
APACHE_HTTP_PORT_NUMBER=80
APACHE_HTTPS_PORT_NUMBER=443
BITNAMI_APP_NAME=moodle
MARIADB_HOST=<cluster endpoint>
MARIADB_PORT_NUMBER=3306
MOODLE_DATABASE_NAME=<a word>
MOODLE_DATABASE_PASSWORD=<a real great password>
MOODLE_DATABASE_USER=<tried existing, root and new usernames>
MOODLE_PASSWORD=<my favorite something>
MOODLE_SITENAME=Yo
MOODLE_USERNAME=<tried the same stuff as database user>
MYSQL_CLIENT_CREATE_DATABASE_PASSWORD=<some new user password>
MYSQL_CLIENT_CREATE_DATABASE_PRIVILEGES=ALL
MYSQL_CLIENT_CREATE_DATABASE_USER=<a new username for ^ password>
```
- Volumes:
aliased as: `moodledata`
container path: `/bitnami` 
Read and Write enabled
- Networking:
	- Bridge 
	- Publicly accessible IP
	- Private IP
	- Ports: `80:80` and `443:443` 
	- Security Group: Same as database cluster
		- (For testing) anything in that security group can access any other member of the same security group, on any port (like the default security group).  That will get more secure after I can get it to run.

**Comment by juan131** [comment]
Hi @adam-phillipps 

Taking a look to the environment variable you setup, it seems you're missing the env. variable **MYSQL_CLIENT_CREATE_DATABASE_NAME**. 

You should also indicate the env. variables **MARIADB_ROOT_USER** and **MARIADB_ROOT_PASSWORD** indicating your RDS database credentials.

Please take a look to the README.md section below:

https://github.com/bitnami/bitnami-docker-moodle/#create-a-database-for-moodle-using-mysql-client

Using that config, Moodle will create its own database during the initialisation on your RDS instance.

**Comment by adam-phillipps** [comment]
Hey @juan131 .  Thanks for the response!  I added those values and am getting the same error.  If I run the containers 2 or more times, I get 1 line of logs that's different.  The difference is that if I run it more than once I can see this (extra line) in the logs: 
`mysql-c INFO ==> Found database <my db name>. Skipping database creation...`

here's a snippet of the logs, up to the fail:

    mysql-c INFO Trying to connect to MySQL server
    mysql-c INFO Found MySQL server listening at moodle-database-cluster.cluster-cfuxgo32bxfw.us-west-2.rds.amazonaws.com:3306
    mysql-c INFO MySQL server listening and working at <my db host>:3306
    **_mysql-c INFO ==> Found database <my db name>. Skipping database creation..._**
    nami INFO mysql-client successfully initialized
    nami INFO Initializing moodle
    mysql-c INFO Trying to connect to MySQL server
    mysql-c INFO Found MySQL server listening at <my db host>:3306
    mysql-c INFO MySQL server listening and working at <my db host>:3306
    Error executing 'postInstallation': Program exited with exit code 1


Maybe I'm just not understanding what the values _should_ be.  Why is there a Maria root user and a Moodle database User and the ability to create a user with the `MYSQL_CLIENT_CREATE_DATABASE_USER` var?  Should the `MYSQL_CLIENT_CREATE_DATABASE_USER` var match the `MOODLE_DATABASE_USER`?  I've tried just about as many permutations on the config as I can think of.  I actually already tried all the config I could get my hands on and whittled it down to the list I sent because changing the other values seemed to have no affect.

Thanks again for the response!  It's no good to be stuck for this long and I have a team waiting on this so I appreciate the help, a lot.

**Comment by juan131** [comment]
Hi @adam-phillipps 

When using more than one container, you need to be careful since every container will consider that it needs to create its database during the initialisation. 

The reason is that we create a flag during the first boot, which is persisted, that allows a container to determine if it should perform a normal initialisation or recover the data from the persisted directory. However, since those containers do not share the persisted volume, once you initialise the container 1, the container 2 doesn't detect that all the initialisation has been performed. In order to solve that, you should use a NFS service to sync the persisted data on every container. In the link below, you'll find a blog that explains how to do it using K8s and WordPress (the process is exactly the same if you use Moodle instead of WP):

https://engineering.bitnami.com/articles/scaling-wordpress-in-kubernetes.html

> Why is there a Maria root user and a Moodle database User and the ability to create a user with the MYSQL_CLIENT_CREATE_DATABASE_USER var? Should the MYSQL_CLIENT_CREATE_DATABASE_USER var match the MOODLE_DATABASE_USER?

- **MYSQL_CLIENT_XXXX** env. vars are used by a MySQL client to prepare the database for Moodle. 
- **MOODLE_DATABASE_XXXX** env. vars used to provide Moodle the credentials to use to connect to the database.

Therefore, the following env. vars should have the same values:

 - **MOODLE_DATABASE_NAME** & **MYSQL_CLIENT_CREATE_DATABASE_NAME**
 - **MOODLE_DATABASE_USER** & **MYSQL_CLIENT_CREATE_DATABASE_USER**
 - **MOODLE_DATABASE_PASSWORD** & **MYSQL_CLIENT_CREATE_DATABASE_PASSWORD**

Then, you have a serie of env. vars used by the MySQL client to access as root user to your database:

- **MARIADB_HOST**
- **MARIADB_PORT_NUMBER**
- **MARIADB_ROOT_USER**
- **MARIADB_ROOT_PASSWORD**

In order to get more debug information on the logs, you can use the env. variable **NAMI_DEBUG** as shown below:

```
- NAMI_DEBUG='--log-level=trace'
```

That might provide you more info about the reason why the initialisation process is failing.

**Comment by adam-phillipps** [comment]
Finally some progress!  Thanks for that.

I'm still hitting some errors, though.  There were a couple that were easy enough to work through but the one that I'm stuck on now is that it seems to be having issues with creating the `mdl_config` table.  

After I turned up verbosity with the `NAMI_DEBUG` var I was able to see this in the logs:

    CREATE TABLE mdl_config (
        id BIGINT(10) NOT NULL auto_increment,
        name VARCHAR(255) COLLATE utf8mb4_general_ci NOT NULL DEFAULT '',
        value LONGTEXT COLLATE utf8mb4_general_ci NOT NULL,
        CONSTRAINT PRIMARY KEY (id),
        UNIQUE KEY mdl_conf_nam_uix (name)
    )
    Error executing 'postInstallation': Default exception handler: DDL sql execution error Debug: Index column size too large. The maximum column size is 767 bytes.

I `docker exec`'d the `admin/install` script and watched its fail message and it mentioned turning `innodb_large_prefix` to `ON` so I did that.  The setting change in the DB does appear to have taken, when I look at it in the DB console but I still see the same error message, in the Moodle logs.

Any idea about that one?

By the way, the verbosity setting was really helpful.  I couldn't find anything for it when I was searching.  Are there any documents where I could find info like that?

Thanks again,
Adam

**Comment by juan131** [comment]
Did you set it globally or did you set it only on a specific database? Did you follow the steps suggested on the link below? 

https://stackoverflow.com/questions/35847015/mysql-change-innodb-large-prefix

> By the way, the verbosity setting was really helpful. I couldn't find anything for it when I was searching. Are there any documents where I could find info like that?

Moodle uses [Nami](https://github.com/bitnami/nami) to initialise the service. You can find information about logging options when using Nami on the link below:

https://github.com/bitnami/nami/blob/master/docs/Nami.md

**Comment by adam-phillipps** [comment]
I didn't follow those steps but I did set the same variables in an RDS way.  These are the ones I've got on there:

    innodb_file_format: Barracuda
    innodb_large_prefix: 'ON'
    general_log: 1
    log_output: TABLE
    slow_query_log: 1

Thanks for the Nami pointer.  I'll take a look at their docs too.

**Comment by juan131** [comment]
I'm not an expert on RDS but, does that configuration applies to a new created database? Moodle will create its own database when setting the env. variables **MYSQL_CLIENT_CREATE_XXX** and I'm not sure if that config is being applied there.

Another alternative you can use is to create the database manually on RDS. E.g. you create a **bitnami_moode** database with all the privileges to a **bn_moodle** identified by some password. Then, when booting your Moodle container, do **NOT** indicate the **MYSQL_CLIENT_CREATE_XXX** env. vars. Just indicate the env. variables: **MOODLE_DATABASE_NAME**, **MOODLE_DATABASE_USER**, **MOODLE_DATABASE_PASSWORD** and **MARIADB_HOST**.

When choosing that configuration, it will assume that you created the database for Moodle and it will populate it without creating it.

**Comment by adam-phillipps** [comment]
Alright, I got it to run.  I had to manually set the character set for the database to utf8.  After that the install script was able to finish and I was able to log in as the admin user.

Thanks again for all your help!

**Comment by juan131** [comment]
Hi @adam-phillipps 

Great!!! Thanks for sharing how you fixed it! I am pretty sure it will help others users facing similar issues. 

I proceed to close this issue but please do not hesitate to reopen it if you consider it necessary.
